hal/nxp_p1021.c

```diff
@@ -1058,26 +1176,421 @@ static void hal_io_init(void)
+} QE_PACKED;
+
+struct qe_firmware {
+    struct qe_header    header;
+    uint8_t             id[62];         /* Null-terminated identifier string */
+    uint8_t             split;          /* 0 = shared I-RAM, 1 = split I-RAM */
+    uint8_t             count;          /* Number of microcode[] structures */
+    struct qe_soc       soc;
+    uint8_t             padding[4];     /* Reserved, for alignment */
+    uint64_t            extended_modes; /* Extended modes */
+    uint32_t            vtraps[8];      /* Virtual trap addresses */
+    uint8_t             reserved[4];    /* Reserved, for future expansion */
+    struct qe_microcode microcode[1];
+    /* All microcode binaries should be located here */
+    /* CRC32 should be located here, after the microcode binaries */
+} QE_PACKED;
+
+static void qe_upload_microcode(const struct qe_firmware *firmware,
+    const struct qe_microcode *ucode)
+{
+    const uint32_t *code = (void*)firmware + ucode->code_offset;
+    unsigned int i;
+
+    wolfBoot_printf("QE: uploading '%s' version %u.%u.%u\n",
+        ucode->id, ucode->major, ucode->minor, ucode->revision);
+
+    /* Use auto-increment */
+    set32(QE_IRAM_IADD, ucode->iram_offset |
+        QE_IRAM_IADD_AIE | QE_IRAM_IADD_BADDR);
+
+    /* Copy 32-bits at a time to iRAM */
+    for (i = 0; i < ucode->count; i++) {
+        set32(QE_IRAM_IDATA, code[i]);
+    }
+}
+
+/* Upload a microcode to the I-RAM at a specific address */
+int qe_upload_firmware(const struct qe_firmware *firmware)
+{
+    unsigned int i, j;
+    uint32_t crc;
+    size_t calc_size = sizeof(struct qe_firmware);
+    size_t length;
+    const struct qe_header *hdr;
+
+    hdr = &firmware->header;
+    length = hdr->length;
+
+    /* Check the magic */
+    if ((hdr->magic[0] != 'Q') || (hdr->magic[1] != 'E') ||
+        (hdr->magic[2] != 'F')) {
+        wolfBoot_printf("QE firmware header invalid!\n");
+        return -1;
+    }
+
+    /* Check the version */
+    if (hdr->version != 1) {
+        wolfBoot_printf("QE version %d unsupported!\n", hdr->version);
+        return -1;
+    }
+
+    /* Validate some of the fields */
+    if ((firmware->count < 1) || (firmware->count > QE_MAX_RISC)) {
+        wolfBoot_printf("QE count %d invalid!\n", firmware->count);
+        return -1;
+    }
+
+    /* Validate the length and check if there's a CRC */
+    calc_size += (firmware->count - 1) * sizeof(struct qe_microcode);
+    for (i = 0; i < firmware->count; i++) {
+        /* For situations where the second RISC uses the same microcode
+         * as the first, the 'code_offset' and 'count' fields will be
+         * zero, so it's okay to add those. */
+        calc_size += sizeof(uint32_t) * firmware->microcode[i].count;
+    }
+
+    /* Validate the length */
+    if (length != calc_size + sizeof(uint32_t)) {
+        wolfBoot_printf("QE length %d invalid!\n", length);
+        return -1;
+    }
+
+#if 0 /* Disabled for wolfBoot */
```

Comment:
Remove dead code or add a config define to enforce CRC checking when appropriate

Reply:
OK if you want to keep this, but please change `#if 0` into something meaninful like `#ifdef QE_CRC_VALIDATE` even if not present in the current config